### PR TITLE
refactor: migrate automl finish_example

### PR DESCRIPTION
**Pull request by bassmang**
Removing finish

### CODE REVIEW

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 1]
```diff
@@ -228,7 +213,7 @@ VW::LEARNER::base_learner* make_automl_with_impl(VW::setup_base_i& stack_builder
                 .set_input_label_type(VW::label_type_t::CB)
                 .set_input_prediction_type(VW::prediction_type_t::ACTION_SCORES)
                 .set_output_label_type(VW::label_type_t::CB)
-                .set_finish_example(::finish_example<config_manager_type>)
+                .set_print_example(print_example_automl)
```
[the remark this answers is not in the document]
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/automl.cc]: split into `set_print_update` and `set_update_stats`

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 2]
```diff
@@ -76,28 +76,28 @@ void persist(automl<CMType>& data, VW::metric_sink& metrics)
 }
 
 template <typename CMType>
-void finish_example(VW::workspace& all, automl<CMType>& data, VW::multi_ex& ec)
+void print_update_automl(VW::workspace& all, VW::shared_data& /* sd */, const automl<CMType>& data,
+    const VW::multi_ex& ec, VW::io::logger& /* unused */)
 {
-  interaction_vec_t* incoming_interactions = ec[0]->interactions;
+  data.adf_learner->print_example(all, ec);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/automl.cc:82]:
```suggestion
  data.adf_learner->print_update(all, ec);
```

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 3]
```diff
@@ -76,28 +76,28 @@ void persist(automl<CMType>& data, VW::metric_sink& metrics)
...
-        [&ec, &incoming_interactions]
-        {
-          for (VW::example* ex : ec) { ex->interactions = incoming_interactions; }
-        });
-
-    data.adf_learner->print_example(all, ec);
-  }
-
-  VW::finish_example(all, ec);
+  VW::scope_exit(
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/reductions/automl.cc:92]: Examples should not be mutated in this function

File: vowpalwabbit/core/src/reductions/automl.cc (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/automl.cc:82]: removing, can use `print` from below reductions instead

File: vowpalwabbit/core/src/reductions/automl.cc (hunk 3 above)
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/automl.cc:92]: none of this seems necessary here

File: vowpalwabbit/core/src/reductions/automl.cc [hunk 4]
```diff
@@ -228,12 +228,13 @@ VW::LEARNER::base_learner* make_automl_with_impl(VW::setup_base_i& stack_builder
                 .set_input_label_type(VW::label_type_t::CB)
                 .set_input_prediction_type(VW::prediction_type_t::ACTION_SCORES)
                 .set_output_label_type(VW::label_type_t::CB)
-                .set_finish_example(::finish_example<config_manager_type>)
-                .set_save_load(save_load_aml<config_manager_type>)
+                .set_print_update(print_update_automl)
+                .set_update_stats(update_stats_automl)
```
[the remark this answers is not in the document]
Comment (bassmang) [reply vowpalwabbit/core/src/reductions/automl.cc:232]: yes (check test 351), does not require any finish logic

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**